### ISSUE TITLE
Rename version to version.sh for compilation with c++20

### ISSUE THREAD

**Issue by Azrael3000** [hypre-space/hypre#513]
We did a compile test recently of our software that includes hypre and if we enabled c++20 it failed because another software package uses  `#include <version>` which is part of the new standard ([0]). Since you have a `version` file [1] this caused the compiler to read the wrong file and fail. It is suggested that you rename the `version` file to `version.sh` for example to solve this issue.

[0] https://en.cppreference.com/w/cpp/header/version
[1] https://github.com/hypre-space/hypre/blob/master/src/utilities/version

**Comment by rfalgout** [comment]
This file isn't part of the hypre installation, so I don't understand why there is a conflict.  You should only have the `include` directory for the hypre installation in your include path when compiling.  Also, you want the angle brackets in `#include <version>` to check system directories first, so you should ensure that you haven't changed this behavior by defining an include path environment variable or something.

**Comment by liruipeng** [comment]
It seems to be related to #506

**Comment by rfalgout** [comment]
Looks like there were similar issues here: https://github.com/spack/spack/issues/11515.  In particular, notice this comment:

  https://github.com/spack/spack/issues/11515#issuecomment-495626063

Finally, it looks like there is a legitimate issue in OpenMPI that is being fixed:

  https://github.com/open-mpi/ompi/issues/9122

From what I see so far, though, this doesn't look like an issue that hypre should try to address, since we don't include "version" (with any case-insensitive spelling) anywhere in our own source files, nor do we install a file of this type for users to accidentally include.

**Comment by aaigner** [comment]
Maybe it depends on the way we build our program:
We reference hypre as a git submodule and add it via cmake `add_subdirectory`. Then we simply link the HYPRE target with our target.
The *utilities* directory that contains the *version* shell script is added to the hypre-target public include directories [here](https://github.com/hypre-space/hypre/blob/7f2762cffbfe0ec3bd2b718ce8baad929874a09f/src/CMakeLists.txt#L383)

I think it is also the combination of compiler and project combination. The used compiler was a non C++20 compiler, thus it has no std-header "version". Therefore, it picks what ever it finds in the include path. Something similar was also discussed in the [boost issue](https://github.com/boostorg/config/issues/345). (which is the source of all evil and unfortunately we can't avoid it)

and... the whole compilation is on Windows which is case-insensitive by the way.

**Comment by rfalgout** [comment]
@osborn9 - Do you have recommendations on how the CMake build should be done here?  User code should only point to the installed version of hypre and not the full source directory.  Thanks!

**Comment by osborn9** [comment]
@aaigner I think using Cmake's ExternalProject (https://cmake.org/cmake/help/latest/module/ExternalProject.html) feature would help. This way you could specify where hypre is built/installed, then you can link to the install location.

**Comment by aaigner** [comment]
@osborn9 Thank you. I know the ExternalProject and we use it for several other third-party projects to set up a proper build environment. For some internal reason it is (or was) the preferred way to add it as a submodule and subdirectory. We may reconsider our decision, though.

Anyway, you may think about renaming/moving the *version*-file since with the words from the boost guys:

> The point is: All of those projects are invalid as of C++20 as (AFAIK) replacing a standard header is UB (undefined behavior) and that is what those projects are doing.

I know in the case of Hypre that is only true for the BUILD_INTERFACE and not the final installed version.

Thank you for your time and comments. Feel free to close the issue if you do not plan any further steps.

**Comment by rfalgout** [comment]
Hi @aaigner .  Just a couple of comments to consider:

* It may be easy for you to just delete the `version` file in hypre as part of your build process, since you are already compiling things in a kind of monolithic manner anyway.
* I disagree with the "invalid" comment from the boost guys for a few reasons.  Even if a project is a C++20 project and also explicitly does `#include <version>` in their own source, as long as the project's own build system ensures that it searches the system include directories first to get the correct `version` file, then this is fine.  Projects can choose to build their libraries in any way they like, as long as it produces a usable installation, which also clearly means that the installation should not have a file named `version` in it.  As far as I can tell, hypre does all of these things correctly.
* The issue here seems to be that you are searching the hypre source directory for include files.  This means that hypre can't have *any* files that clash with file names you include in your source, because you might get our file instead of your own.  It's really important to point only to the installed hypre `include` directory for this reason.  We put the `HYPRE_` prefix in front of our include files to provide some protection from this (we still have a few things to fix here, though).

Related to the above, have you tried to compile/install hypre by itself to see if this issue persists.  Maybe there is something more subtle going on that I don't understand.  Let me know, and sorry for pushing back so much.  :)

**Comment by aaigner** [comment]
Hi!
Sorry for the late answer. Rather busy times these days ...

First of all. No worries I never take this personal. It's an issue discussion board and it's fine to have different opinions. Worst case - you proof me wrong and I learn something new. :)

* :+1: I checked the installed version of hypre and as expected the `version` file is not installed. Thus, there should not be any problem during compilation.
* Don't worry, we will find a workflow for our build routine. Either deleting/renaming files, extracting hypre as a cmake-externalproject,... or just simply using a C++20 compiler which provides a std-version header.

And for me the issue is the workflow that cmake allows to use: By adding projects via `add_subdirectory` you share the same build interface with all its pros and cons.
I know some projects using this approach for internal and external libraries, but maybe the intended usage was for internal libraries only. Thus, you control the build environment.

So for me: I am not sure, if we did it wrong or it is really bad practice to have a "version" file without file ending in an include directory (even if its only the build interface). To avoid any problems in the future I will try to avoid "version" files as far as possible.

**Comment by adayton1** [comment]
Hi, we have also encountered this issue. The problem is that I have a windows build that doesn't use CMake (and it would be difficult to change to using CMake, especially since my project has about 50 dependencies). Is there a technical reason you can't change the name of the script? Or move it to a scripts subdirectory?

**Comment by rfalgout** [comment]
Hi @adayton1 ,

I think it makes more sense to fix the fundamental problem: you should not point to the hypre source tree as part of your compile.  This is a general rule for compiling code with multiple library dependencies.  It isn't a cmake issue specifically.  Regardless of the approach you use to build your code, you should only point to the hypre installation (the `lib` and `include` directories).  Is there a reason why you can't change the include directory you use for hypre to just be `<path-to-hypre-installation>/include`?

Sorry for pushing back, but I think it's important to promote scalable approaches for building code.  Today the issue is a file named `version`, but if you are pointing to the hypre source, then it could be any file name we have in hypre now or choose to create in the future vs any include file name any code or compiler standard chooses to use.  By limiting this interaction to the installation directories and protecting the file names with package name prefixes (`HYPRE_` in our case), this greatly improves interoperability, maintainability, etc.  There are a few files in our install directory that are not currently protected (it's on our to do list), so if you have any problems with these, we will be more than happy to fix it.

If I am misunderstanding something here, please let me know.  Thanks!

**Comment by dbarbier** [comment]
Hello, just hit this issue too.  I will modify our CMakeLists.txt to remove this file as part of our build, but I would like to comment about

>  Even if a project is a C++20 project and also explicitly does #include <version> in their own source, as long as the project's own build system ensures that it searches the system include directories first to get the correct version file, then this is fine

In fact that may not be that easy.  Here is a simplistic example:
```
echo '#include "mylibrary.h"' > mylibrary.cxx
echo '#include <some_external_library.h>' > mylibrary.h
echo | sudo tee /usr/include/some_external_library.h
```
mylibrary.cxx uses quotes and angle brackets as advertised.  It compiles fine:
```
$ g++ -c  -I. -I[...]/hypre/src/utilities  mylibrary.cxx
$
```
But if some_external_library.h contains `#include <version>`, it breaks
```
$ echo '#include <version>' | sudo tee /usr/include/some_external_library.h
$ g++ -c  -I. -I[...]/hypre/src/utilities  mylibrary.cxx
In file included from /usr/include/some_external_library.h:1,
                 from mylibrary.h:1,
                 from mylibrary.cxx:1:
[...]/hypre/src/utilities/version:1:2: error: invalid preprocessing directive #!
    1 | #!/bin/sh
      |  ^
[...]/hypre/src/utilities/version:2:3: error: invalid preprocessing directive #Copyright
    2 | # Copyright 1998-2019 Lawrence Livermore National Security, LLC and other
      |   ^~~~~~~~~
```

**Comment by pbauman** [comment]
So this issue is arising directly in HYPRE itself (with a forthcoming version of ROCm) in a way similar to as @dbarbier describes.

<pre>
hipcc  -O2 -x hip -std=c++14 -I/home/pabauman/work/libs/rocprim/install/include -I/home/pabauman/work/libs/rocthrust/install/include  -DHAVE_CONFIG_H -I.. -I./.. -I./../struct_mv -I.   -I/opt/rocm/include      -I/usr/include/x86_64-linux-gnu/mpich  -c device_utils.c -o device_utils.obj
In file included from device_utils.c:10:
In file included from ./_hypre_utilities.hpp:131:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/execution_policy.h:33:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/cpp/execution_policy.h:64:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/cpp/detail/sort.h:22:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/detail/sequential/sort.h:62:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/detail/sequential/sort.inl:23:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/detail/sequential/stable_merge_sort.h:58:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/detail/sequential/stable_merge_sort.inl:22:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/merge.h:677:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/detail/merge.inl:27:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/detail/adl/merge.h:44:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/hip/detail/merge.h:39:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/extrema.h:800:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/detail/extrema.inl:22:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/detail/generic/extrema.h:87:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/detail/generic/extrema.inl:29:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/reduce.h:781:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/detail/reduce.inl:28:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/detail/generic/reduce_by_key.h:87:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/detail/generic/reduce_by_key.inl:37:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/scan.h:1656:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/detail/scan.inl:28:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/detail/adl/scan.h:44:
In file included from /home/pabauman/work/libs/rocthrust/install/include/thrust/system/hip/detail/scan.h:48:
In file included from /home/pabauman/work/libs/rocprim/install/include/rocprim/rocprim.hpp:33:
In file included from /home/pabauman/work/libs/rocprim/install/include/rocprim/intrinsics.hpp:29:
In file included from /home/pabauman/work/libs/rocprim/install/include/rocprim/intrinsics/thread.hpp:27:
In file included from /home/pabauman/work/libs/rocprim/install/include/rocprim/intrinsics/../detail/various.hpp:38:
./version:1:2: error: invalid preprocessing directive
#!/bin/sh
 ^
</pre>

If I look at that header I can see the `#include <version>`. And even if I force an include path where the C++ `version` header is located, the clang compiler is still choosing the one in the current directory first. So I am at a loss as to how to avoid this failure in HYPRE without renaming the `version` file to something else (#725  was just a suggestion).

**Comment by rfalgout** [comment]
I believe there is a way to fix this without changing the `version` file name, but at this point I think we will go ahead and change the name anyway.  My biggest complaint above was about user build systems searching the hypre source directory for include files, instead of searching the install directory.

FYI, the alternative way to fix this would be to embed the hypre directory structure in the include lines (e.g., change `#include "_hypre_utilities.h"` to `#include "utilities/_hypre_utilities.h"`) and remove everything except the top-level `src` directory from the include file search path in each `Makefile` (not sure what the equivalent approach would be on the CMake side).  Presumably the compiler would never search the `src/utilities` directory in that case, and it would not find the `utilities/version` file by mistake.  A long time ago, this was exactly how we implemented header file inclusion in hypre, but the downside to this approach is that any directory structure changes require more source modifications than with the search path approach, so we switched over to the current approach instead.
